src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+{
+    fd = bp_sys::open(path, O_RDONLY | flags, mode);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool istream::open(const char *path, const int flags, const mode_t mode)
+{
+    bool r = open_nx(path, flags, mode);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::close_nx() noexcept
+{
+    if (bp_sys::close(fd) < 0) {
+        io_error = errno;
+        return false; // Failed
+    }
+
+    fd = -1;
+    return true;
+}
+
+bool istream::close()
+{
+    bool r = close_nx();
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+int istream::current_state() noexcept
+{
+    int bits = 0;
+    if (eof_state) bits |= io_states::eofbit;
+    if (!buf) bits |= io_states::buffer_failbit;
+    if (string_failed) bits |= io_states::string_failbit;
+    if (io_error) bits |= io_states::io_failbit;
+    return bits;
+}
+
+bool istream::good() noexcept
+{
+    return (current_state() == 0);
+}
+
+bool istream::eof() noexcept
+{
+    return eof_state;
+}
+
+bool istream::buffer_failure() noexcept
+{
+    return (!buf);
+}
+
+bool istream::string_failure() noexcept
+{
+    return string_failed;
+}
+
+int istream::io_failure() noexcept
+{
+    return io_error;
+}
+
+void istream::clear() noexcept
+{
+    eof_state = false;
+    string_failed = false;
+    io_error = 0;
+}
+
+int istream::getc_nx() noexcept
+{
+    if (eof() || buffer_failure() || io_failure()) {
+        // We don't care about string_failure() in getc() function.
+        return -1;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(1);
+        if (r == 0) {
+            eof_state = true;
+            return -1;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    char result = *buf->get_ptr(0);
+    buf->consume(1);
+    return result;
+}
+
+int istream::getc()
+{
+    int r = getc_nx();
+    if (r < 0) {
+        throw_exception_on(io_states::eofbit | io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::get_line_nx(std::string &dest, char delim) noexcept
+{
+    if (!good()) {
+        return false;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(buf->get_size());
+        if (r == 0) {
+            eof_state = true;
+            return false;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return false;
+        }
+    }
+    dest.clear();
+
+    char *ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    char *eptr = static_cast<char *>(std::memchr(ptr, delim, len));
+
+    if (eptr == ptr) {
+        buf->consume(1);
+        return true; // An empty line
+    }
+    while (eptr == nullptr) {
+        // If it hasn't any delim; We try to read as much as possible to reach the EOF or find a
```

Comment:
`// while no delimiter is found, append the entire buffer contents to the 'dest' string and refill the buffer`